src/Storages/ExternalStream/Kafka/KafkaSink.cpp

```diff
@@ -210,26 +223,72 @@ KafkaSink::KafkaSink(
+/// However, it's still possible that, one single row is still too big and it exceeds that limit. There is nothing we can do about it for now.
+void KafkaSink::addMessageToBatch(char * pos, size_t len, size_t total_len)
 {
-    StringRef key = message_key_expr ? keys_for_current_batch[current_batch_row++] : "";
-
-    /// Data at pos (which is in the WriteBuffer) will be overwritten, thus it must be kept somewhere else (in `batch_payload`).
-    nlog::ByteVector payload {len};
-    payload.resize(len); /// set the size to the right value
-    memcpy(payload.data(), pos, len);
-
-    current_batch.push_back(rd_kafka_message_t{
-        .partition = next_partition,
-        .payload = payload.data(),
-        .len = len,
-        .key = const_cast<char *>(key.data),
-        .key_len = key.size,
-        ._private = this,
-    });
+    auto pending_size = pending_data.size();
+
+    /// There are complete data to consume.
+    if (len > 0)
+    {
+        StringRef key = message_key_expr ? keys_for_current_batch[current_batch_row++] : "";
+
+        /// Data at pos (which is in the WriteBuffer) will be overwritten, thus it must be kept somewhere else (in `batch_payload`).
+        auto msg_size = pending_size + len;
+        nlog::ByteVector payload{msg_size};
+        payload.resize(msg_size); /// set the size to the right value
+        if (pending_size)
+            memcpy(payload.data(), pending_data.data(), pending_size);
+        memcpy(payload.data() + pending_size, pos, len);
+
+        current_batch.push_back(rd_kafka_message_t{
+            .partition = next_partition,
+            .payload = payload.data(),
+            .len = msg_size,
+            .key = const_cast<char *>(key.data),
+            .key_len = key.size,
+            ._private = this,
+        });
+
+        batch_payload.push_back(std::move(payload));
+        ++state.outstandings;
+
+        pending_data.resize(0);
+        pending_size = 0;
+        rows_in_current_message = 0;
+    }
+
+    if (len == total_len)
+        /// Nothing left
+        return;
 
-    batch_payload.push_back(std::move(payload));
-    ++state.outstandings;
+    /// There are some remaining incomplete data, copy them to pending_data.
+    auto remaining = total_len - len;
+    pending_data.resize(pending_size + remaining);
+    memcpy(pending_data.data() + pending_size, pos + len, remaining);
+
+    external_stream_counter->addToMessagesBySize(1);
+}
+
+void KafkaSink::tryCarryOverPendingData()
+{
+    /// If there are pending data and it can be fit into the buffer, then write the data back to the buffer,
+    /// so that we can use the buffer to limit the message size.
+    /// If the pending data are too big, that means we get a over-size row.
+    if (!pending_data.empty() && pending_data.size() < wb->available())
+    {
+        wb->write(pending_data.data(), pending_data.size());
```

Comment:
In general, this is just a potential optimization which we can do in next PR.

Reply:
I have considered using another buffer, for example, the `pending_data` for batching data for a message. It doesn't look it's a better solution. Because it has to copy the data from buffer to pending_data for every single row.

Another way to optimize this is to swap pending_data and the buffer of `wb`. I will try this in next PR.